sql/src/main/java/io/crate/analyze/AnalyzedTableElements.java

```diff
@@ -143,11 +156,26 @@ private void expandColumn(AnalyzedColumnDefinition column) {
                 if (column.hasNotNullConstraint() && !primaryKeys().contains(fqn)) { // Columns part of pk are implicitly not null
                     notNullColumns.add(fqn);
                 }
+                addNotNullFromChildren(column);
             }
         }
         return notNullColumns;
     }
 
+    /**
+     * Recursively add all not null constraints from child columns (object columns)
+     */
+    private void addNotNullFromChildren(AnalyzedColumnDefinition column) {
+        if (column.hasChildren()) {
+            for (AnalyzedColumnDefinition child : column.children()) {
+                if (child.hasNotNullConstraint()) {
+                    notNullColumns.add(child.ident().fqn());
+                }
+                addNotNullFromChildren(child);
```

Comment:
I wouldn't use recursion for something that's easily expressed with a `while` loop (as it's a bit slower due to lack of tail recursion)